src/libsyntax/ext/tt/transcribe.rs

```diff
@@ -176,9 +176,34 @@ pub fn transcribe(cx: &ExtCtxt,
                 stack.push(Frame::Delimited { forest: delimited, idx: 0, span: span });
                 result_stack.push(mem::replace(&mut result, Vec::new()));
             }
-            quoted::TokenTree::Token(sp, tok) => {
-                let mut marker = Marker(cx.current_expansion.mark);
-                result.push(noop_fold_tt(TokenTree::Token(sp, tok), &mut marker).into())
+            quoted::TokenTree::Token(sp, tok, escape_hygiene) => {
+                let result_tok = match tok {
+                    token::Ident(ident) |
+                    token::Lifetime(ident) => {
+                        let sp_ctxt = if escape_hygiene { cx.call_site() } else { sp }.ctxt();
+                        let sp = sp.with_ctxt(sp_ctxt.apply_mark(cx.current_expansion.mark));
+
+                        let ident_ctxt = if escape_hygiene {
+                            cx.call_site().ctxt()
+                        } else {
+                            ident.ctxt.apply_mark(cx.current_expansion.mark)
+                        };
+                        let ident = Ident { ctxt: ident_ctxt, ..ident };
+
+                        TokenTree::Token(sp, match tok {
```

Comment:
The context of this span needs to match the context of the identifier or lifetime token for proc-macros to be coherent (a proc-macro identifer token is just an interned string and a span; when we reconstruct the `libsyntax::TokenTree` the interned string turns into an identifier with the span's context).

You might have chosen to use `cx.current_expansion.mark` for better diagnostics (even if the ident resolves at call site, it should show up in diagnostics at the call site). To achieve this, the plan is for spans to have separate `SyntaxContext`s for hygiene and for diagnostics. I'm working on a PR for this now.

Reply:
Yeah, I was in a bit of a conundrum over this... I'm glad you're working on separating out hygiene and diagnostic contexts. To be clear, how should the span be correctly constructed under this new model?

I guess we can wait for your PR to land, before merging this, presuming it won't be too much more work? Do post a link here when you open it.